### PR TITLE
tECDSA signing: Properly map DKG result to a wallet signing group

### PR DESCRIPTION
**Pull request by lukasz-zimnoch**
Refs: #3042 
Depends on:  #3120

The final signing group established at the end of the DKG protocol may differ from the original group outputted by the sortition protocol. The final group should consist from members who behaved correctly during the DKG protocol and final members indexes should reflect the position of particular members in that group in a correct way. This is not only important for code consistency but is also required by the `WalletRegistry` contract where functions like `isWalletMember` rely on properly built wallet members hash.

This changeset applies several changes that adjust the code to the aforementioned requirements and improves the documentation, logging, and test coverage around it.

### CODE REVIEW

File: pkg/tbtc/node_test.go [hunk 1]
```diff
@@ -388,3 +389,96 @@ func TestDkgRetryLoop(t *testing.T) {
 		})
 	}
 }
+
+func TestFinalSigningGroup(t *testing.T) {
+	beaconConfig := &ChainConfig{
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply pkg/tbtc/node_test.go]: See 98e45904

File: pkg/tbtc/node.go [hunk 2]
```diff
@@ -280,32 +283,54 @@ func (n *node) joinDKGIfEligible(seed *big.Int, startBlockNumber uint64) {
...
-					operatingMemberIndexes,
-				)
+				// Final signing group may differ from the original DKG
+				// group outputted by the sortition protocol. One need to
+				// determine the final signing group based on the selected
+				// group members who behaved correctly during DKG protocol.
+				finalSigningGroupOperators, finalSigningGroupMembersIndexes, err :=
+					finalSigningGroup(
+						selectedSigningGroupOperators,
+						operatingMemberIndexes,
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply pkg/tbtc/node.go:302]: Not sure if I understood correctly but `decideSigningGroupMemberFate` always takes the `misbehavedMembersIndices` from the published result and uses that slice to update our `operatingMemberIndexes` so we always use the published view of properly operating members. Doesn't it solve the problem? Everything else should be handled during result approval where we should check whether the published result match our expectations.

**Review by tomaszslabon — APPROVED** [review]

File: pkg/tbtc/node.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (pdyraga) [reply pkg/tbtc/node.go:302]: Okay, I see I missed the fact `operatingMemberIndexes` value - which is originally set to `result.Group.OperatingMemberIDs()` - is overwritten by `decideSigningGroupMemberFate`. 

This looks good but I think it's worth to unit test the interaction of `decideSigningGroupMemberFate` and `finalSigningGroup`. If it is hard given how the code is currently organized, I would not mind calling these two functions separately from a unit test.

I do not consider it blocking for this PR, maybe we can address it in #3209?

**Review by pdyraga — APPROVED** [review]

File: pkg/tbtc/node.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply pkg/tbtc/node.go:302]: I covered `decideSigningGroupMemberFate` with unit tests within #3209 (see https://github.com/keep-network/keep-core/pull/3209/commits/e2d1d20753c042c4456c51b89edce3ab43d0addf). I think there is no sense to write a case where those functions will be called separately from a single test. This way we will not test their real interaction that occurs within `node.joinDKGIfEligible`. If something changes there, the test will not catch this. We should stress this with an integration test of `node.go` instead (there is a TODO for `node.go` tests that should be addressed in the future). Second, both `decideSigningGroupMemberFate` and `finalSigningGroup` are now covered with unit tests separately so we are sure that if they receive correct parameters, they return a correct result.

**Pull request merged**